NE-367: Add logLevel and operatorLogLevel APIs for DNS (#931)
openshift/enhancements: enhancements/dns/dns-operator-operand-logging-level.md

```go
type DNSSpec struct {
	// <snip>

	// operatorLogLevel controls the logging level of the DNS Operator.
	// See LogLevel for more information about each available logging level.
	//
	// +optional
	OperatorLogLevel LogLevel `json:"operatorLogLevel"`
}
```

This new field would allow a cluster administrator to specify the desired logging level specifically for the DNS Operator.

Additionally, a new `LogLevel` of type [LogLevel](https://github.com/openshift/api/blob/master/operator/v1/types.go#L62) will be added for CoredDNS logging :

```go
// We will be enabling the CLASSES(https://github.com/coredns/coredns/tree/master/plugin/log#syntax) of coredns w.r.t to the LogLevel we have defined in openshift api.

// So,

// logLevel "Normal"  will enable errors class as we enable errors plugin by default log . { class error }

// logLevel "Debug" will enable log . { class denial error }

// logLevel "Trace" will enable  log . { class all }
type DNSSpec struct {
	
	// logLevel describes the logging verbosity of the DNSController for CoreDNS.
	//
	// +optional

> 	LogLevel LogLevel `json:"logLevel"`

Reviewer: add defaulting to the standard loglevel.

Author: sorry @sttts forgot to update here as had added it the api https://github.com/openshift/api/pull/1031
@sttts just a request if would be able to review https://github.com/openshift/api/pull/1031 then that will be great.

Author: @sttts PTAL just added the changes you requested. Thanks in advance!

## Summary

This enhancement describes the API and code changes necessary to expose
a means to change the DNS Operator and CoreDNS Logging Levels to
cluster administrators.

## Motivation

Supporting a trivial way to raise the verbosity of the DNS Operator and it's Operands (CoreDNS) would make debugging
cluster dns operator and CoreDNS issues easier for cluster administrators and OpenShift developers.

To have the CoreDNS enable the following classes such as error, denial and all.
denial: either NXDOMAIN or nodata responses (Name exists, type does not). A nodata response sets the return code to NOERROR.
error: SERVFAIL, NOTIMP, REFUSED, etc. Anything that indicates the remote server is not willing to resolve the request.
all: the default - nothing is specified. Using of this class means that all messages will be logged whatever we mix together with "all".

Also, a logging level API for the DNS Operator would assist OpenShift developers working on the DNS Operator who may
desire more in-depth logging statements when working on the operator's controllers.

Additionally, a logging level API for CoreDNS logs would assist cluster administrators who wish to have more control

> over the output of their DNSController's CoreDNS logs.

Reviewer 2: This section would flow better if you swapped these two paragraphs (adjusting transitions as necessary).

### Goals

> * Add a user-facing API for controlling the run-time verbosity of the [OpenShift DNS Operator and CoreDNS](https://github.com/openshift/cluster-dns-operator)

Reviewer 2: ```suggestion
* Add a user-facing API for controlling the run-time verbosity of the [OpenShift DNS Operator and CoreDNS](https://github.com/openshift/cluster-dns-operator).
```

### Non-Goals

* Change the default logging verbosity of the DNS Operator or the CoreDNS in production OCP clusters.

## Proposal

### DNS Operator Log Level API
We will be creating an API for **field** `operatorLogLevel` in DNSSpec in accordance with [LogLevel](https://github.com/openshift/api/blob/master/operator/v1/types.go#L71) type.
Valid values will be : "Normal", "Debug", "Trace", "TraceAll".
We will using [logrus](https://github.com/sirupsen/logrus#level-logging) to enable the log levels for the operator level logging.
Logrus has seven logging levels: Trace, Debug, Info, Warning, Error, Fatal and Panic.

> log.Trace("Something very low level.")
> log.Debug("Useful debugging information.")
> log.Info("Something noteworthy happened!")
> log.Warn("You should probably take a look at this.")
> log.Error("Something failed but I'm not quitting.") // Calls os.Exit(1) after logging log.Fatal("Bye.") // Calls panic() after logging log.Panic("I'm bailing.")

Reviewer 2: ````suggestion
```go
log.Trace("Something very low level.")
log.Debug("Useful debugging information.")
log.Info("Something noteworthy happened!")
log.Warn("You should probably take a look at this.")
log.Error("Something failed but I'm not quitting.")

// Calls os.Exit(1) after logging:
log.Fatal("Bye.")

// Calls panic() after logging:
log.Panic("I'm bailing.")
```
````

After the logging level on a Logger is set, log entries with that severity or anything above it will be logged.

> Eg- Default log.SetLevel(log.InfoLevel) will log anything that is info or above (warn, error, fatal, panic).

Reviewer 2: ```suggestion
For example, `log.SetLevel(log.InfoLevel)` will log anything that is info or above (warn, error, fatal, panic).  This is the default log level.  
```

> So, we will be setting operatorLogLevel in a separate controller to watch dnses and set log level.
> operatorLogLevel: "Normal" will set logrus.SetLogLevel("Info")
> operatorLogLevel: "Debug" will set logrus.SetLogLevel("Debug")
> operatorLogLevel: "Trace" will set  logrus.SetLogLevel("Trace")

Reviewer 2: ```suggestion
So, we will be reading `operatorLogLevel` in a separate controller to watch dnses and setting log level.
`operatorLogLevel: "Normal"` will set `logrus.SetLogLevel("Info")`.
`operatorLogLevel: "Debug"` will set `logrus.SetLogLevel("Debug")`.
`operatorLogLevel: "Trace"` will set  `logrus.SetLogLevel("Trace")`.
```

### CoreDNS Log Level API

Valid values for logLevel are: "Normal", "Debug", "Trace", "TraceAll" as per [Operator Api](https://github.com/openshift/api/blob/master/operator/v1/types.go#L62)

> We will be enabling the CLASSES(https://github.com/coredns/coredns/tree/master/plugin/log#syntax) of coredns w.r.t to the logLevel we have defined in openshift api.

Reviewer 2: ```suggestion
We will enable logging of CoreDNS's [classes of responses](https://github.com/coredns/coredns/tree/master/plugin/log#syntax) that correspond to the log level specified in the API.
```

So,

> logLevel "Normal"  will enable errors class as we enable errors plugin by default log . { class error }
> 
> logLevel "Debug" will enable log . { class denial error }
> 
> logLevel "Trace" will enable  log . { class all }

Reviewer 2: What is the relationship between the `errors` plugin and the `log . { class error }` stanza?  I think we discussed this before—doesn't `errors` log things like connection errors whereas `log . { class error }` logs error responses?
```suggestion
`logLevel: "Normal"`  will enable the "errors" class: `log . { class error }`.

`logLevel: "Debug"` will enable the "denial" class: `log . { class denial error }`.

`logLevel: "Trace"` will enable the "all" class: `log . { class all }`.

Note that the `errors` plugin is always enabled.  The `errors` plugin logs TCP/UDP connection errors whereas `log . { class error }` logs DNS error responses (such as SERVFAIL).  
```

This would be a good place to add a note that CoreDNS reloads its configuration without requiring a restart, so the operator can adjust CoreDNS's log level just by updating the Corefile configmap without need to restart the pod.

We will creating an API under DNSSpec with **field name** `operatorlogLevel` of type [LogLevel](https://github.com/openshift/api/blob/master/operator/v1/types.go#L71)

```go
type DNSSpec struct {
	// <snip>

	// operatorLogLevel controls the logging level of the DNS Operator.
	// See LogLevel for more information about each available logging level.
	//
	// +optional
	OperatorLogLevel LogLevel `json:"operatorLogLevel"`
}
```

This new field would allow a cluster administrator to specify the desired logging level specifically for the DNS Operator.

> Additionally, a new `LogLevel` of type [LogLevel](https://github.com/openshift/api/blob/master/operator/v1/types.go#L62) will be added for CoredDNS logging :

Reviewer 2: ```suggestion
Additionally, a new `LogLevel` field of type [LogLevel](https://github.com/openshift/api/blob/master/operator/v1/types.go#L62) will be added for CoredDNS logging:
```

```go
// We will be enabling the CLASSES(https://github.com/coredns/coredns/tree/master/plugin/log#syntax) of coredns w.r.t to the LogLevel we have defined in openshift api.

// So,

// logLevel "Normal"  will enable errors class as we enable errors plugin by default log . { class error }

// logLevel "Debug" will enable log . { class denial error }

// logLevel "Trace" will enable  log . { class all }
type DNSSpec struct {
	
	// logLevel describes the logging verbosity of the DNSController for CoreDNS.
	//
	// +optional
	LogLevel LogLevel `json:"logLevel"`
}

```

> Both of these new APIs would be accompanied by appropriate `LogLevel` definitions:

Reviewer 2: ```suggestion
Both of these new APIs would use the existing `LogLevel` definitions:
```

	Debug LogLevel = "Debug"

	// Trace is used when something went really badly and even more verbose logs are needed.  Logging every function call as part of a common operation, to tracing execution of a query.  In kube, this is probably glog=6.
	Trace LogLevel = "Trace"

	// TraceAll is used when something is broken at the level of API content/decoding.  It will dump complete body content.  If you turn this on in a production cluster
	// prepare from serious performance issues and massive amounts of logs.  In kube, this is probably glog=8.
	TraceAll LogLevel = "TraceAll"
)
```

### User Stories

* As an OpenShift Cluster Administrator, I want to be able to raise the logging level of the DNS Operator and CoreDNS so that I can more quickly
track down OpenShift DNS issues.

* Some users recently added the new alert 'CoreDNS is returning SERVFAIL for X% of requests alert' to the recent updates of OCP.
  Adding this prometheus alert is nice, but it would be more useful we can see which request are getting SERVFAIL response.
  So we would to enable the log plugin for CoreDNS to log queries.
  

> * Some user want to avoid use of tcpdump to see the queries and want log plugin to be enabled to log queries in coredns.

Reviewer 2: These items seem more like motivations, and they all go back to the same user story: I want to change the log level of CoreDNS so that I can diagnose DNS issues.  Maybe you could simplify by saying that we have the alert, but diagnosing it sometimes requires using tcpdump whereas being able to adjust CoreDNS's log level would suffice in many cases.  This could also be moved to the "motivation" section (although I won't press it if you prefer to keep it here).  

Something that could be added as a user story is that some users actually want logs of every DNS query for auditing purposes, similar to access logs for ingress or audit logs for the API.

Reviewer 2: I believe my previous comment here still applies.

### Implementation Details/Notes/Constraints [optional]

### Risks and Mitigations

> Raising the logging verbosity for any component typically results in larger log files that grow quickly.

Reviewer 2: If possible, each risk should have a mitigation.  For example, the godoc for the new API field could include a warning that setting `logLevel: Trace` will produce extremely verbose logs.

Author: @Miciah Do you mean add a comment in the API PR and recreate the go docs

Reviewer 2: No, I mean the point of having a "Risks **and Mitigations**" section is that you'll describe the mitigation for each risk.

## Design Details

### Open Questions [optional]
N/A

### Test Plan

Unit tests will be written to test if setting LogLevel sets the respective logging in CoreDNS.
Unit tests will be written to test if setting operatorLogLevel sets the respective logging in DNS Operator.

### Graduation Criteria

N/A

#### Dev Preview -> Tech Preview

N/A

#### Tech Preview -> GA

N/A

#### Removing a deprecated feature

N/A

### Upgrade / Downgrade Strategy

On downgrade, any logging options are ignored by the DNS Operator and CoreDNS.

> A harmless logging level configmap in the `openshift-dns` namespace may be left behind.

Reviewer 2: Why would there be an extra configmap?

Author: I think I need to remove this.  the downgraded operator will update the configmap and delete the log stanzas.

---
title: dns-operator-operand-logging-level
authors:
  - "@miheer"
reviewers:
  - "@alebedev87"
  - "@candita"
  - "@frobware"
  - "@knobunc"
  - "@Miciah"
  - "@rfredette"
approvers:
  - "@frobware"
  - "@knobunc"
  - "@Miciah"
  - "@alebedev87"
  - "@rfredette"
  - "@candita"
creation-date: 2021-10-14
last-updated: 2021-10-14
status: implementable
---

# DNS Log Level API

## Release Signoff Checklist

- [X] Enhancement is `implementable`
- [ ] Design details are appropriately documented from clear requirements
- [ ] Test plan is defined
- [ ] Operational readiness criteria is defined
- [ ] Graduation criteria for dev preview, tech preview, GA
- [ ] User-facing documentation is created in [openshift-docs](https://github.com/openshift/openshift-docs/)

## Summary

This enhancement describes the API and code changes necessary to expose
a means to change the DNS Operator and CoreDNS Logging Levels to
cluster administrators.

## Motivation

Supporting a trivial way to raise the verbosity of the DNS Operator and its Operands (CoreDNS) would make debugging
the Operator and CoreDNS issues easier for cluster administrators and OpenShift developers.

> For logging purposes, CoreDNS defines several classes of responses, such as error, denial and all.
> denial: either NXDOMAIN or nodata responses (Name exists, type does not). A nodata response sets the return code to NOERROR.
> error: SERVFAIL, NOTIMP, REFUSED, etc. Anything that indicates the remote server is not willing to resolve the request.
> all: all responses, including successful responses, errors, and denials.

Reviewer 2: Please fix the formatting.
```suggestion
For logging purposes, CoreDNS defines several classes of responses:
* denial: NXDOMAIN and nodata responses (name exists, record type does not). A nodata response sets the return code to NOERROR.
* error: SERVFAIL, NOTIMP, REFUSED, etc. Anything that indicates the remote server is not willing to resolve the request.
* all: all responses, including successful responses, errors, and denials.
```

A logging level API for CoreDNS logs would assist cluster administrators who wish to have more control
over CoreDNS logs.

Also, a logging level API for the DNS Operator would assist OpenShift developers working on the DNS Operator who may
desire more in-depth logging statements when working on the operator's controllers.

### Goals

Add a user-facing API for controlling the run-time verbosity of the [OpenShift DNS Operator and CoreDNS](https://github.com/openshift/cluster-dns-operator).
### Non-Goals

* Change the default logging verbosity of the DNS Operator or CoreDNS in production OCP clusters.

## Proposal

### DNS Operator Log Level API

> We will be creating an API for **field** `operatorLogLevel` in DNSSpec in accordance with `DNSLogLevel` type.

Reviewer 2: Why the emphasis?  You ended up defining a new type, right?  
```suggestion
 We will be defining a new API field `operatorLogLevel` in `DNSSpec` with newly defined type `DNSLogLevel`.
 This type is similar to the existing `LogLevel` type except that the values of `DNSLogLevel` are a subset of the values of `LogLevel`.  
```

> Valid values will be : "Normal", "Debug", "Trace".

Reviewer 2: ```suggestion
Valid values will be the following: "Normal", "Debug", "Trace".
```

`operatorLogLevel: "Normal"` will set `logrus.SetLogLevel("Info")`.

`operatorLogLevel: "Debug"` will set `logrus.SetLogLevel("Debug")`.

`operatorLogLevel: "Trace"` will set  `logrus.SetLogLevel("Trace")`.

### CoreDNS Log Level API

Valid values for coredns logLevel are: "Normal", "Debug", "Trace" as per `DNSLogLevel` type.

We will enable logging of CoreDNS's [classes of responses](https://github.com/coredns/coredns/tree/master/plugin/log#syntax) that correspond to the log level specified in the API.
So,

`logLevel: "Normal"`  will enable the "errors" class: `log . { class error }`.

`logLevel: "Debug"` will enable the "denial" class: `log . { class denial error }`.

`logLevel: "Trace"` will enable the "all" class: `log . { class all }`.

Note that the `errors` plugin is always enabled.  The `errors` plugin logs TCP/UDP connection errors whereas `log . { class error }` logs DNS error responses (such as SERVFAIL).
The CoreDNS reloads its configuration without requiring a restart, so the operator can adjust CoreDNS's log level just by updating the Corefile configmap without need to restart the pod.

> We will be adding an API field `operatorlogLevel` in DNSSpec for the type DNSLogLevel

Reviewer 2: ```suggestion
 We will be adding an API field `operatorlogLevel` in `DNSSpec` with the type `DNSLogLevel`:
```

```go
// operatorLogLevel controls the logging level of the DNS Operator.
// Valid values are: "Normal", "Debug", "Trace".
// Defaults to "Normal".
// setting operatorLogLevel: Trace will produce extremely verbose logs.
// +optional
// +kubebuilder:default=Normal
OperatorLogLevel DNSLogLevel `json:"operatorLogLevel,omitempty"`
```
This new field would allow a cluster administrator to specify the desired logging level specifically for the DNS Operator.

> Additionally, a new `LogLevel` of type `DNSLogLevel` will be added for CoredDNS logging :

Reviewer 2: ```suggestion
 Additionally, a new API field `LogLevel` of type `DNSLogLevel` will be added to specify the log level for CoreDNS:
```

```go
// logLevel describes the desired logging verbosity for CoreDNS.
// Any one of the following values may be specified:
// * Normal logs errors from upstream resolvers.
// * Debug logs errors, NXDOMAIN responses, and NODATA responses.
// * Trace logs errors and all responses.
//  Setting logLevel: Trace will produce extremely verbose logs.
// Valid values are: "Normal", "Debug", "Trace".
// Defaults to "Normal".
// +optional
// +kubebuilder:default=Normal
LogLevel DNSLogLevel `json:"logLevel,omitempty"`
```

> Both of these new APIs would use the exisiting `DNSLogLevel` definitions:

Reviewer 2: ```suggestion
Both of these new API fields use the aforementioned `DNSLogLevel` type, which is defined as follows:
```

```go

// +kubebuilder:validation:Enum:=Normal;Debug;Trace
type DNSLogLevel string

var (
// Normal is the default.  Normal, working log information, everything is fine, but helpful notices for auditing or common operations.  In kube, this is probably glog=2.
DNSLogLevelNormal DNSLogLevel = "Normal"

// Debug is used when something went wrong.  Even common operations may be logged, and less helpful but more quantity of notices.  In kube, this is probably glog=4.
DNSLogLevelDebug DNSLogLevel = "Debug"

// Trace is used when something went really badly and even more verbose logs are needed.  Logging every function call as part of a common operation, to tracing execution of a query.  In kube, this is probably glog=6.
DNSLogLevelTrace DNSLogLevel = "Trace"
)

```

### User Stories

* As an OpenShift Cluster Administrator, I want to be able to raise the logging level of the DNS Operator and CoreDNS so that I can more quickly
track down OpenShift DNS issues.

> * Some users recently added the new alert 'CoreDNS is returning SERVFAIL for X% of requests alert' to the recent updates of OCP.

Reviewer 2: What do you mean by "Some users" added the alert?  Is this referring to an alert that is in the product?

Author: no some customers add prometheus alerts so they will need this logging to be set

>   Adding this prometheus alert is nice, but it would be more useful we can see which request are getting SERVFAIL response.
>   So we would to enable the log plugin for CoreDNS to log queries.

Reviewer 2: ```suggestion
Adding this Prometheus alert is useful, but it would be more useful if we could see which requests were getting SERVFAIL responses.
So we would like to configure the log plugin for CoreDNS to log queries.
```

### Implementation Details/Notes/Constraints [optional]

### Risks and Mitigations

Raising the logging verbosity for any component typically results in larger log files that grow quickly.

## Design Details

### Open Questions [optional]
N/A

### Test Plan

Unit tests will be written to test if setting LogLevel sets the respective logging in CoreDNS.
Unit tests will be written to test if setting operatorLogLevel sets the respective logging in DNS Operator.

### Graduation Criteria

N/A

#### Dev Preview -> Tech Preview

N/A

#### Tech Preview -> GA

N/A

#### Removing a deprecated feature

N/A

### Upgrade / Downgrade Strategy

On downgrade, any logging options are ignored by the DNS Operator and CoreDNS.
The downgraded operator will update the configmap and delete the log stanzas.

### Version Skew Strategy

N/A

## Implementation History

[API Implementation](https://github.com/openshift/api/pull/1031/commits/b606c856dae1b68383f0d068f78ca0823c82ad03/)

## Drawbacks

## Alternatives

* Don't provide any DNS logging level APIs for the operator and coredns (current behavior)

> * Raise current verbosity of the DNS Operator and coredns (not desirable)

Reviewer 2: A third alternative is tcpdump.  

Normally we'd want a little more discussion as to why the alternative was not considered.

### API Extensions

### Operational Aspects of API Extensions

#### Failure Modes

> #### Support Procedures

Reviewer 2: These should be filled out.

- [ ] Test plan is defined
- [ ] Operational readiness criteria is defined
- [ ] Graduation criteria for dev preview, tech preview, GA
- [ ] User-facing documentation is created in [openshift-docs](https://github.com/openshift/openshift-docs/)

## Summary

This enhancement describes the API and code changes necessary to expose
a means to change the DNS Operator and CoreDNS Logging Levels to
cluster administrators.

## Motivation

* As an OpenShift Cluster Administrator, I want to be able to raise the logging level of the DNS Operator and CoreDNS so that I can more quickly
  track down OpenShift DNS issues.

* Supporting a trivial way to raise the verbosity of the DNS Operator and its Operands (CoreDNS) would make debugging
the Operator and CoreDNS issues easier for cluster administrators and OpenShift developers.

For logging purposes, CoreDNS defines several classes of responses, such as error, denial and all.
* denial: either NXDOMAIN or nodata responses (Name exists, type does not). A nodata response sets the return code to NOERROR.
* error: SERVFAIL, NOTIMP, REFUSED, etc. Anything that indicates the remote server is not willing to resolve the request.

> * all: all responses, including successful responses, errors, and denials.
> A logging level API for CoreDNS logs would assist cluster administrators who wish to have more control
> over CoreDNS logs.

Reviewer 2: Needs a linebreak to prevent lines 53-54 from being formatted as part of the bullet on line 52.  
```suggestion
* all: all responses, including successful responses, errors, and denials.

A logging level API for CoreDNS logs would assist cluster administrators who wish to have more control
over CoreDNS logs.
```

Also, a logging level API for the DNS Operator would assist OpenShift developers working on the DNS Operator who may
desire more in-depth logging statements when working on the operator's controllers.

* Some users want to add new prometheus alert 'CoreDNS is returning SERVFAIL for X% of requests alert' to the recent updates of OCP.
  Adding this Prometheus alert is useful, but it would be more useful if we could see which requests were getting SERVFAIL responses.

>   So we would like to configure the log plugin for CoreDNS to log queries.

Reviewer 2: I'm still confused as to what the use-case is here, or how this relates to the existing alert: https://github.com/openshift/cluster-dns-operator/blob/531e38a2f640bfbefb7126d6e701afad8db6f911/manifests/0000_90_dns-operator_03_prometheusrules.yaml#L32-L43

## Proposal

### DNS Operator Log Level API
We will be defining a new API field `operatorLogLevel` in `DNSSpec` with newly defined type `DNSLogLevel`.
This type is similar to the existing `LogLevel` type except that the values of `DNSLogLevel` are a subset of the values of `LogLevel`.
Valid values will be the following: "Normal", "Debug", "Trace".
We will use [logrus](https://github.com/sirupsen/logrus#level-logging) to set the log level for the operator-level logging.
Logrus has seven logging levels: Trace, Debug, Info, Warning, Error, Fatal and Panic.
```go
log.Trace("Something very low level.")

log.Debug("Useful debugging information.")

log.Info("Something noteworthy happened!")

log.Warn("You should probably take a look at this.")

log.Error("Something failed but I'm not quitting.")

// Calls os.Exit(1) after logging:
log.Fatal("Bye.")

// Calls panic() after logging:
log.Panic("I'm bailing.")
```
After the logging level on a Logger is set, log entries with that severity or anything above it will be logged.
For example, `log.SetLevel(log.InfoLevel)` will log anything that is info or above (warn, error, fatal, panic).  This is the default log level.  

> So, we will be reading `operatorLogLevel` in a separate controller to watch dnses and setting log level.

Reviewer 2: You didn't end up using a separate controller, so this text should be updated when you get the chance.

// * Trace logs errors and all responses.
//  Setting logLevel: Trace will produce extremely verbose logs.
// Valid values are: "Normal", "Debug", "Trace".
// Defaults to "Normal".
// +optional
// +kubebuilder:default=Normal
LogLevel DNSLogLevel `json:"logLevel,omitempty"`
```

Both of these new API fields use the aforementioned `DNSLogLevel` type, which is defined as follows:
```go

// +kubebuilder:validation:Enum:=Normal;Debug;Trace
type DNSLogLevel string

var (
// Normal is the default.  Normal, working log information, everything is fine, but helpful notices for auditing or common operations.  In kube, this is probably glog=2.
DNSLogLevelNormal DNSLogLevel = "Normal"

// Debug is used when something went wrong.  Even common operations may be logged, and less helpful but more quantity of notices.  In kube, this is probably glog=4.
DNSLogLevelDebug DNSLogLevel = "Debug"

// Trace is used when something went really badly and even more verbose logs are needed.  Logging every function call as part of a common operation, to tracing execution of a query.  In kube, this is probably glog=6.
DNSLogLevelTrace DNSLogLevel = "Trace"
)

> ```

Reviewer 2: The definitions we ended up with in https://github.com/openshift/api/pull/1031 look a bit different; the enhancement text should be updated eventually.
